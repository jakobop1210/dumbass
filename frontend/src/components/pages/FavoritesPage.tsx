import { useSelector } from 'react-redux'
import { RootState } from '../../redux/store'
import { Artist, Song } from '../../lib/types'
import { ClipLoader } from 'react-spinners'
import Breadcrumb from '../atoms/Breadcrumb'
import CardView from '../organisms/CardView'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import { useFavorites } from '../../hooks/useFavorites'

export default function FavoritesPage() {
    const username = useSelector((state: RootState) => state.user.username)

    const { dataSongs, loading, error, dataArtists } = useFavorites(username)

    // Processing and mapping data for Card components
    const artists: Artist[] = dataArtists?.getArtistsByIds || []
    const songs: Song[] = dataSongs?.getSongsByIds || []

    const artistCardData = artists.map(
        (artist: Artist): ArtistCardProps => ({
            cardType: 'artist',
            id: artist.id,
            title: artist.name,
            alternateNames: artist.alternate_names,
            imageUrl: artist.image_url,
            rating: artist.average_rating,
            numOfRatings: artist.number_of_ratings
        })
    )

    const songCardData = songs.map(
        (song: Song): SongCardProps => ({
            cardType: 'song',
            id: song.id,
            title: song.title,
            artist: song.artist_names,
            imageUrl: song.header_image_url,
            rating: song.average_rating,
            numOfRatings: song.number_of_ratings,
            releaseDate: song.release_date
        })
    )

    if (!username) {
        return (
            <div className='h-96 flex items-center justify-center text-blueGray italic'>
                <h1>You need to be logged in to add favorites...</h1>
            </div>
        )
    }

    return (
        <main className='w-full flex flex-col'>
            <Breadcrumb items={[{ name: 'Favorites' }]} />
            <section className='w-full flex flex-col justify-center items-center mt-5'>
                {loading ? (
                    <div className='h-96 flex items-center'>
                        <ClipLoader color={'#8fc0a9'} size={100} />
                    </div>
                ) : error ? (
                    <div className='h-96 flex items-center'>
                        <h1 className='text-blueGray text-2xl'>
                            Error loading data :/
                        </h1>
                    </div>
                ) : (
                    <>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Songs'
                                cardData={songCardData}
                                customErrorMessage='You have no favorite songs...'
                            />
                        </section>
                        <section className='w-full flex justify-center'>
                            <CardView
                                title='Favorited Artists'
                                cardData={artistCardData}
                                customErrorMessage='You have no favorite artists...'
                            />
                        </section>
                    </>
                )}
            </section>
        </main>
    )
}
