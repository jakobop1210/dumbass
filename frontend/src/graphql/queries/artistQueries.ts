import { gql } from '@apollo/client'

export const GET_TOP_ARTISTS = gql`
    query ($limit: Int) {
        getTopArtists(limit: $limit) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTIST_BY_ID = gql`
    query ($id: ID!) {
        getArtistById(id: $id) {
            alternate_names
            description
            image_url
            name
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTISTS_BY_ID = gql`
    query ($ids: [ID!]!) {
        getArtistsByIds(ids: $ids) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`
