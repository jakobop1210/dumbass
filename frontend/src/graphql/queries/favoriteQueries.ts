import { gql } from '@apollo/client'

/**
 * GraphQL query: CHECK_IF_FAVORITE.
 *
 * This query checks if a specific item (song or artist) is marked as a favorite by a user.
 * It requires the username of the user, the type of the item (song or artist), and the item's ID.
 *
 * @param {string} $username - The username of the user.
 * @param {string} $type - The type of the item ('song' or 'artist').
 * @param {number} $targetId - The ID of the item (song or artist).
 * @returns {boolean} A boolean indicating whether the item is a favorite of the user.
 */
export const CHECK_IF_FAVORITE = gql`
    query ($username: String!, $type: String!, $targetId: Int!) {
        checkIfFavorite(username: $username, type: $type, targetId: $targetId)
    }
`

/**
 * GraphQL query: GET_FAVORITES.
 *
 * This query retrieves a list of favorite items (songs or artists) for a specific user.
 * It requires the username of the user and returns an array of objects, each representing a favorite item.
 * Each object includes the type of the item (song or artist) and its ID.
 *
 * @param {string} $username - The username of the user whose favorites are being requested.
 * @returns {Array<Object>} An array of objects, each containing the type ('song' or 'artist') and the targetId of the favorite item.
 */
export const GET_FAVORITES = gql`
    query ($username: String!) {
        getFavorites(username: $username) {
            type
            targetId
        }
    }
`
