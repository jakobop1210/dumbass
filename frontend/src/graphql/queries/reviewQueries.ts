import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_REVIEWS_BY_TARGET_ID.
 *
 * This query fetches reviews for a specific target item, identified by its type and ID.
 * The target can be either a song or an artist. The query returns a list of reviews, 
 * each containing the username of the reviewer, the content of the review, and the rating given.
 *
 * @param {string} $targetType - The type of the target for which reviews are being fetched ('song' or 'artist').
 * @param {number} $targetId - The unique identifier of the target (song or artist).
 * @returns {Array<Object>} An array of review objects, each containing the userName, content, and rating of the review.
 */
export const GET_REVIEWS_BY_TARGET_ID = gql`
    query ($targetType: String!, $targetId: Int!) {
        getReviewsByTarget(targetType: $targetType, targetId: $targetId) {
            userName
            content
            rating
        }
    }
`
