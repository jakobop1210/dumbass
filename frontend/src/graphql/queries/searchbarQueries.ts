import { gql } from '@apollo/client'

export const SEARCHBAR_DROPDOWN = gql`
    query ($searchType: String!, $searchString: String!, $limit: Int) {
        searchSearchbar(
            searchType: $searchType
            searchString: $searchString
            limit: $limit
        ) {
            ... on Artist {
                name
                id
            }
            ... on Song {
                title
                id
            }
        }
    }
`
