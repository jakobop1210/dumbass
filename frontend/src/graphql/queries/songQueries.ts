import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_SONG_BY_ID.
 *
 * This query fetches detailed information about a song by its unique ID. 
 * The details include the song's lyrics, header image URL, release date, title, artist names, average rating, and the number of ratings.
 *
 * @param {string} $id - The unique ID of the song to fetch.
 * @returns {Object} An object containing detailed information about the song.
 */
export const GET_SONG_BY_ID = gql`
    query ($id: ID!) {
        getSongById(id: $id) {
            lyrics
            header_image_url
            release_date
            title
            artist_names
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_BY_ID.
 *
 * This query retrieves detailed information for a list of songs based on their unique IDs.
 * The details include the song's ID, title, artist names, header image URL, release date, primary artist ID, average rating, and number of ratings.
 *
 * @param {Array<number>} $ids - An array of IDs of the songs to fetch.
 * @returns {Array<Object>} An array of objects, each containing detailed information about a song.
 */
export const GET_SONGS_BY_ID = gql`
    query ($ids: [Int]!) {
        getSongsByIds(ids: $ids) {
            id
            title
            artist_names
            header_image_url
            release_date
            primary_artist_id
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_ON_TITLE.
 *
 * This query fetches a list of songs based on a search term, with optional parameters for pagination and sorting.
 * The details returned for each song include the header image URL, ID, title, artist names, average rating, number of ratings, and release date.
 *
 * @param {string} [$title] - The title or part of the title of the songs to search for.
 * @param {number} [$limit] - The maximum number of songs to return.
 * @param {string} [$sort] - The sorting criteria (e.g., 'rating', 'alphabetical').
 * @param {number} [$page] - The page number for pagination.
 * @returns {Array<Object>} An array of objects, each containing information about a song.
 */
export const GET_SONGS_ON_TITLE = gql`
    query ($title: String, $limit: Int, $sort: String, $page: Int) {
        getSongsOnTitle(
            title: $title
            limit: $limit
            sort: $sort
            page: $page
        ) {
            header_image_url
            id
            title
            artist_names
            average_rating
            number_of_ratings
            release_date
        }
    }
`

/**
 * GraphQL query: COUNT_SONGS.
 *
 * This query counts the number of songs in the database that match a given title or search term.
 *
 * @param {string} [$title] - The title or part of the title of the songs to count.
 * @returns {number} The count of songs matching the search criteria.
 */
export const COUNT_SONGS = gql`
    query ($title: String) {
        countSongs(title: $title)
    }
`
