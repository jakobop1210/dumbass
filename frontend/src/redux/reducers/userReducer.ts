import { createReducer } from '@reduxjs/toolkit';
import { setUserLogin, setUserLogout, setUserName } from '../actions/userActions';

interface UserState {
    loggedIn: boolean;
    username: string | null;
}

const initialState: UserState = {
    loggedIn: false,
    username: null,
};

const userReducer = createReducer(initialState, (builder) => {
    builder
        .addCase(setUserLogin, (state) => {
            state.loggedIn = true;
        })
        .addCase(setUserLogout, (state) => {
            state.loggedIn = false;
            state.username = null; // Nullify username
        })
        .addCase(setUserName, (state, action) => {
            state.username = action.payload; // Set username
        });
});

export default userReducer;
