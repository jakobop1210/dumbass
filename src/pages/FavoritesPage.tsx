import { useState } from 'react'
import CommonDropdown from '../components/atoms/CommonDropdown'
import Paginate from 'react-paginate'
import {
    ArtistCardProps,
    SongCardProps
} from '../components/molecules/ArtistSongCard'
import CardView from '../components/views/CardView'
import SortIcon from '@mui/icons-material/Sort'
import GradeIcon from '@mui/icons-material/Grade'
import SortByAlphaIcon from '@mui/icons-material/SortByAlpha'
import FilterAltIcon from '@mui/icons-material/FilterAlt'
import PersonIcon from '@mui/icons-material/Person'
import MusicNoteIcon from '@mui/icons-material/MusicNote'

/**
 * @component FavoritesPage
 * 
 * `FavoritesPage` is a React functional component that provides an interface for users 
 * to interact with their favorite songs and artists.
 */
export default function FavoritesPage() {
    const FiftycentProps = {
        cardType: 'artist',
        imageUrl:
            'https://www.uka.no/uploads/cache/66/e7/66e75771d31a087bd8754021b203d98c.jpg',
        title: '50 Cent',
        alternateNames: ['Fiddy', 'Boo Boo'],
        rating: 4.5,
        numOfRatings: 1000000
    } as ArtistCardProps

    const InDaClubProps = {
        cardType: 'song',
        imageUrl:
            'https://upload.wikimedia.org/wikipedia/en/1/12/50_Cent_-_In_Da_Club_-_CD_cover.jpg',
        title: 'In Da Club',
        artist: '50 Cent',
        rating: 4.5,
        numOfRatings: 1000000,
        releaseDate: '2003-01-07'
    } as SongCardProps

    const allData = Array(10)
        .fill(InDaClubProps)
        .concat(Array(10).fill(FiftycentProps))

    const [currentPage, setCurrentPage] = useState(0)
    const itemsPerPage = 12

    const handlePageClick = (data: { selected: number }) => {
        setCurrentPage(data.selected)
    }

    const offset = currentPage * itemsPerPage
    const currentData = allData.slice(offset, offset + itemsPerPage)

    return (
        <div className='w-full'>
            <div className='flex justify-center gap-10 my-10'>
                <CommonDropdown
                    label='Sort by'
                    icon={<SortIcon />}
                    filterOptions={['Rating', 'Alphabetical']}
                    optionIcons={[<GradeIcon />, <SortByAlphaIcon />]}
                />
                <CommonDropdown
                    label='Filter by'
                    icon={<FilterAltIcon />}
                    filterOptions={['Artists', 'Songs']}
                    optionIcons={[<PersonIcon />, <MusicNoteIcon />]}
                />
            </div>
            <div className='w-full flex flex-col justify-center items-center'>
                <CardView cardData={currentData} />
                <Paginate
                    previousLabel={'Previous'}
                    nextLabel={'Next'}
                    breakLabel={'...'}
                    breakClassName={'break-me'}
                    pageCount={Math.ceil(allData.length / itemsPerPage)}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={handlePageClick}
                    containerClassName={
                        'pagination flex justify-center space-x-2 mb-5'
                    }
                    activeClassName={
                        'active bg-blue-200 flex items-center justify-center'
                    }
                    pageClassName={
                        'text-black border rounded px-3 py-2 hover:bg-blue-300 flex items-center justify-center'
                    }
                    pageLinkClassName={
                        'w-full h-full flex items-center justify-center'
                    }
                    previousClassName={
                        'text-black border rounded px-3 py-2 hover:bg-blue-300 flex items-center justify-center'
                    }
                    nextClassName={
                        'text-black border rounded px-3 py-2 hover:bg-blue-300 flex items-center justify-center'
                    }
                    previousLinkClassName={
                        'w-full h-full flex items-center justify-center'
                    }
                    nextLinkClassName={
                        'w-full h-full flex items-center justify-center'
                    }
                    disabledClassName={
                        'text-gray-300 border rounded px-3 py-2 opacity-50 cursor-not-allowed flex items-center justify-center'
                    }
                    disabledLinkClassName={
                        'opacity-50 cursor-not-allowed w-full h-full flex items-center justify-center'
                    }
                    breakLinkClassName={
                        'border-b border-black w-full h-full flex items-center justify-center'
                    }
                />
            </div>
        </div>
    )
}
