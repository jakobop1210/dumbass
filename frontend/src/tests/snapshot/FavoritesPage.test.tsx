import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import { MockedProvider } from '@apollo/client/testing'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import userReducer from '../../redux/reducers/userReducer'
import FavoritesPage from '../../components/pages/FavoritesPage'
import { GET_FAVORITES } from '../../graphql/queries/favoriteQueries'
import { GET_SONGS_BY_ID } from '../../graphql/queries/songQueries'
import { GET_ARTISTS_BY_ID } from '../../graphql/queries/artistQueries'

const mockStore = configureStore({
    reducer: {
        user: userReducer
    },
    preloadedState: {
        user: { loggedIn: true, username: 'testUser' }
    }
})

const mocks = [
    {
        request: {
            query: GET_FAVORITES,
            variables: { username: 'testUser' }
        },
        result: {
            data: {
                getFavorites: [
                    { type: 'song', targetId: 1 },
                    { type: 'artist', targetId: 2 }
                ]
            }
        }
    },
    {
        request: {
            query: GET_SONGS_BY_ID,
            variables: { ids: [1] }
        },
        result: {
            data: {
                getSongsByIds: [
                    {
                        id: 39,
                        title: 'Fireman',
                        artist_names: 'Lil Wayne',
                        header_image_url:
                            'https://images.rapgenius.com/82b537c1008d3b1a3abb77c51a21a141.600x600x1.jpg',
                        release_date: '2005-10-25',
                        primary_artist_id: '4',
                        average_rating: 4.5,
                        number_of_ratings: 2
                    }
                ]
            }
        }
    },
    {
        request: {
            query: GET_ARTISTS_BY_ID,
            variables: { ids: [2] }
        },
        result: {
            data: {
                getArtistsByIds: [
                    {
                        id: 1,
                        name: 'Cam’ron',
                        alternate_names: [
                            "Killa Cam'",
                            'Cameron Giles',
                            'Cameron Ezike Giles',
                            'Cameron E. Giles'
                        ],
                        description: [
                            "Cameron “Cam'ron” Giles is perhaps the quintessential Rap Genius artist. Killa Cam started out alongside Big L, Ma$e, his cousin Bloodshed (R.I.P.) and others in the short-lived group Children of the Corn.  He had several successful solo albums, but it’s his 2004 record Purple Haze that stands as the quintessence of all things Cam'ronian.",
                            'Since then, Cam has kept his computers ‘puting on a series of albums and mixtapes by himself, with Dipset, and with his standout new protege Vado.  We here at RG salute Mr. Giles, without whom this site might not exist.  May your diamonds keep on doing the Macarena.'
                        ],
                        image_url:
                            'https://images.genius.com/bbfe1c5fc61df408e120c06bd8e423d8.613x613x1.jpg',
                        average_rating: 4,
                        number_of_ratings: 2,
                        __typename: 'Artist'
                    }
                ]
            }
        }
    }
]

test('FavoritesPage snapshot', () => {
    const { asFragment } = render(
        <Provider store={mockStore}>
            <MockedProvider mocks={mocks} addTypename={false}>
                <Router>
                    <FavoritesPage />
                </Router>
            </MockedProvider>
        </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})
