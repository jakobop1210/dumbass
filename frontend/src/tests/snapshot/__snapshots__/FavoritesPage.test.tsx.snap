// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoritesPage snapshot 1`] = `
<DocumentFragment>
  <main
    class="w-full flex flex-col"
  >
    <div
      class="w-1/2 text-left absolute"
    >
      <nav
        aria-label="breadcrumb"
        class="w-max"
      >
        <ol
          class="flex w-full flex-wrap items-center rounded-md bg-opacity-60 py-2 px-4 "
        >
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <a
              class="opacity-60 duration-300 hover:text-green"
              href="/"
            >
              <svg
                class="h-4 w-4"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
                />
              </svg>
            </a>
            <span
              class="pointer-events-none mx-2 select-none font-sans text-sm font-normal leading-normal antialiased"
            >
              /
            </span>
          </li>
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <div
              class="opacity-70"
            >
              <span>
                Favorites
              </span>
            </div>
          </li>
        </ol>
      </nav>
    </div>
    <section
      class="w-full flex flex-col justify-center items-center mt-5"
    >
      <div
        class="h-96 flex items-center"
      >
        <span
          style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #8fc0a9; border-bottom-color: transparent; border-left-color: #8fc0a9; border-right-color: #8fc0a9; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
        />
      </div>
    </section>
  </main>
</DocumentFragment>
`;
